Simplify attribute type decoration

kamipo opened the pull request:
Using `decorate_matching_attribute_type` is overkill to do that.

kamipo committed:
Simplify attribute type decoration

Using `decorate_matching_attribute_type` is overkill to do that.

eugeneius commented:
`define_attribute` runs three times: when the [schema loads](https://github.com/rails/rails/blob/dcf545d20ca6110a360780bda1ba058d7083167a/activerecord/lib/active_record/model_schema.rb#L510), when [attributes are defined](https://github.com/rails/rails/blob/dcf545d20ca6110a360780bda1ba058d7083167a/activerecord/lib/active_record/attributes.rb#L259), and when [attributes are decorated](https://github.com/rails/rails/blob/dcf545d20ca6110a360780bda1ba058d7083167a/activerecord/lib/active_record/attribute_decorators.rb#L53). That means these type decorations will now be applied at least twice, and potentially three times since user-defined attributes can also be decorations (https://github.com/rails/rails/pull/39820, https://github.com/rails/rails/pull/39830). I don't think the additional runtime overhead is worth the simplification here.

eugeneius commented:
We could extract a method to override during the attribute decoration phase instead:

```patch
diff --git a/activerecord/lib/active_record/attribute_decorators.rb b/activerecord/lib/active_record/attribute_decorators.rb
index 0b66043d2a..32aed3329d 100644
--- a/activerecord/lib/active_record/attribute_decorators.rb
+++ b/activerecord/lib/active_record/attribute_decorators.rb
@@ -49,10 +49,14 @@ def decorate_matching_attribute_types(matcher, decorator_name, &block)
         def load_schema!
           super
           attribute_types.each do |name, type|
-            decorated_type = attribute_type_decorations.apply(name, type)
+            decorated_type = apply_attribute_type_decorations(name, type)
             define_attribute(name, decorated_type)
           end
         end
+
+        def apply_attribute_type_decorations(name, type)
+          attribute_type_decorations.apply(name, type)
+        end
     end

     class TypeDecorator # :nodoc:
diff --git a/activerecord/lib/active_record/attribute_methods/time_zone_conversion.rb b/activerecord/lib/active_record/attribute_methods/time_zone_conversion.rb
index f97412b1cd..5a294305e7 100644
--- a/activerecord/lib/active_record/attribute_methods/time_zone_conversion.rb
+++ b/activerecord/lib/active_record/attribute_methods/time_zone_conversion.rb
@@ -64,7 +64,7 @@ def map_avoiding_infinite_recursion(value)
       end

       module ClassMethods # :nodoc:
-        def define_attribute(name, cast_type, **)
+        def apply_attribute_type_decorations(name, cast_type)
           if create_time_zone_conversion_attribute?(name, cast_type)
             cast_type = TimeZoneConverter.new(cast_type)
           end
diff --git a/activerecord/lib/active_record/locking/optimistic.rb b/activerecord/lib/active_record/locking/optimistic.rb
index 8770909086..80082f4575 100644
--- a/activerecord/lib/active_record/locking/optimistic.rb
+++ b/activerecord/lib/active_record/locking/optimistic.rb
@@ -165,7 +165,7 @@ def update_counters(id, counters)
             super
           end

-          def define_attribute(name, cast_type, **) # :nodoc:
+          def apply_attribute_type_decorations(name, cast_type) # :nodoc:
             if lock_optimistically && name == locking_column
               cast_type = LockingType.new(cast_type)
             end
```

kamipo committed:
Remove internal attribute decoration code

The internal attribute decoration is almost covered by the `attribute`
with block, we don't need to have much code and extra class attribute.

kamipo commented:
The purpose of the simplification is to remove much extra code, extra class attribute, and extra hook point, I've extracted the first commit as an easier review part (I supposed no one disagree the smaller part).

If you prefer the extraction, I can pick the diff first before the removing.

kamipo commented:
As the code (`attribute_types.each do ...`) shows, originally all registered decoration matchers are invoked for all defined attributes, so unless having user-defined attributes on the database columns, so the number of invocations of decoration matchers does not change.

https://github.com/rails/rails/blob/dcf545d20ca6110a360780bda1ba058d7083167a/activerecord/lib/active_record/attribute_decorators.rb#L49-L55

Let me know if still have any concerns.

eugeneius commented:
My concern isn't how often the decoration matcher runs, but how many times the type is decorated.

```patch
diff --git a/activerecord/test/cases/attributes_test.rb b/activerecord/test/cases/attributes_test.rb
index 15475d12d8..14f2ddd4d6 100644
--- a/activerecord/test/cases/attributes_test.rb
+++ b/activerecord/test/cases/attributes_test.rb
@@ -87,6 +87,7 @@ class CustomPropertiesTest < ActiveRecord::TestCase
           attribute :starts_at, :datetime, precision: 3, default: -> { Time.now.utc }
         end

+        assert_instance_of ActiveSupport::Type::DateTime, klass.type_for_attribute(:starts_at).__getobj__
         assert_instance_of ActiveSupport::TimeWithZone, klass.new.starts_at
       end
     end
```

```
Failure:
ActiveRecord::CustomPropertiesTest#test_time_zone_aware_attribute [/Users/eugene/src/rails/activerecord/test/cases/attributes_test.rb:90]:
Expected #<ActiveRecord::Type::DateTime:0x00007ff3d8df1640 @precision=3, @scale=nil, @limit=nil> to be an instance of ActiveRecord::Type::DateTime, not ActiveRecord::AttributeMethods::TimeZoneConversion::TimeZoneConverter.
```

The second commit resolves ☝️, but the same problem remains for user-defined timezone attributes that don't specify a type, which was the motivating example for https://github.com/rails/rails/pull/39830. Do you not consider this double-typecasting at runtime to be a problem?

eugeneius commented:
I agree that the `AttributeDecorators` abstraction is too heavyweight for the small number of places it's actually used, by the way. I do think this is a good change, I'd just prefer not to regress runtime performance at all if possible.

kamipo commented:
> The second commit resolves ☝️, but the same problem remains for user-defined timezone attributes that don't specify a type, which was the motivating example for #39830. Do you not consider this double-typecasting at runtime to be a problem?

I've get your point. I'll address that in a later commit.

> I agree that the AttributeDecorators abstraction is too heavyweight for the small number of places it's actually used, by the way. I do think this is a good change, I'd just prefer not to regress runtime performance at all if possible.

As I said above, the purpose of the simplification is to remove much extra code, extra class attribute, and extra hook point.

About runtime performance, as long as the number of invocations of decoration matchers is the same (i.e. not having user-defined attributes on the database columns), removing the extra `load_schema!` hook point (it avoids extra re-`define_attribute`, and changes slower block matcher invocation to just a method call) obviously makes lesser overhead than before.

I am missing something about the performance?

https://github.com/rails/rails/blob/dcf545d20ca6110a360780bda1ba058d7083167a/activerecord/lib/active_record/attribute_decorators.rb#L49-L55

eugeneius commented:
Yes, removing that stuff will definitely make defining attributes faster. By "runtime performance" I meant when the application is serving requests, rather than when it's initialising; I was thinking of the overhead of the double-typecasting.

kamipo committed:
Avoid double time zone converter decoration when user-defined timestamp attribute with implicit type

If type is given as a symbol, it will lookup a fresh type object via
`Type.lookup`, but if type is omitted, `type_for_attribute` will lookup
the database type which is already decorated by `define_attribute`.

To avoid the double decoration, skip extra decoration if its type is
already decorated.

kamipo commented:
I've addressed the double decoration e0366022f6651a0e53af690394fb1957d3a5bcbc, it no longer have runtime overhead.

eugeneius commented on activerecord/lib/active_record/attribute_methods/serialization.rb:
```diff
@@ -69,12 +69,19 @@ def serialize(attr_name, class_name_or_coder = Object)
+          type, options = attributes_to_define_after_schema_loads[attr_name]
+
+          attribute(attr_name) do |cast_type|
+            if type && !type.is_a?(Proc)
+              cast_type = _lookup_cast_type(attr_name, type, options)
+            end
```
It looks like this `type.is_a?(Proc)` check is here to avoid decorating with `Type::Serialized` twice, but there are now multiple ways that the type could be specified as a proc. How about checking the type directly:

```suggestion
            cast_type = _lookup_cast_type(attr_name, type, options) if type
            cast_type = cast_type.subtype if cast_type.is_a?(Type::Serialized)
```

kamipo replied:
Originally there is no official way to decorate with `Type::Serialized` twice.

Personally I'd like to remove this part entirely, but existing tests expects `serialize` can overlay on the user-defined attribute. This part is for passing the tests.

https://github.com/rails/rails/blob/57daae230b9dc3f5166e78e0e6d103bc8faadd24/activerecord/test/cases/json_attribute_test.rb#L13-L14

https://github.com/rails/rails/blob/57daae230b9dc3f5166e78e0e6d103bc8faadd24/activerecord/test/cases/json_shared_test_cases.rb#L237-L239

kamipo replied:
> Personally I'd like to remove this part entirely, but existing tests expects `serialize` can overlay on the user-defined attribute. This part is for passing the tests.

In other words, we need to allow `serialize` on user-defined attribute with symbol for passing tests, but I'd not like to allow `serialize` on user-defined attribute with block which is internally used feature.

eugeneius replied:
Right, my suggestion is another way to make those tests pass. I find `cast_type.is_a?(Type::Serialized)` easier to understand than `type.is_a?(Proc)`.

But it's just a preference, there's no practical benefit. I think I've delayed this PR enough. Thanks for your patience 🙇
